Add Mat.DataPtr methods for direct access to OpenCV data

vcabbage opened the pull request:
@deadprogram It doesn't seem necessary to return an error for the single byte variants. The int8 version probably isn't that useful, but I could imagine a scenario where someone wants to access the raw bytes even if the Mat is of a different type.

Resolves #250

deadprogram commented on core_test.go:
```diff
@@ -213,6 +213,99 @@ func TestMatToBytes(t *testing.T) {
 	}
 }
 
+func TestMatDataPtr(t *testing.T) {
```
Seems to me like it would be a good idea to have code in these tests to set/get a few values in the Mat.

deadprogram replied:
Looking good now, thanks for adding the additional tests.

vcabbage commented:
Added the get/set tests, which made me realize that I was using `Mat_ToBytes` improperly since it allocates and copies the bytes in C++. I added a new `Mat_DataPtr` function to avoid that. I don't normally write C++, so I'd appreciate a double check of that code/any suggestions to make it better.

I think `Mat_DataPtr` could be used for `Mat.ToBytes` as well so that the data data isn't copied twice (once in C++ and once in Go), but I didn't want to expand the scope of this change.

deadprogram commented on core.cpp:
```diff
@@ -66,6 +66,10 @@ struct ByteArray Mat_ToBytes(Mat m) {
     return toByteArray(reinterpret_cast<const char*>(m->data), m->total() * m->elemSize());
 }
 
+struct DataPtr Mat_DataPtr(Mat m) {
```
Seems like we could reuse ByteArray struct here instead of creating a new type. Basically just do what you already implemented but with ByteArray, and since Mat_DataPtr does not perform the memcpy, it does the right thing.

deadprogram commented on core.h:
```diff
@@ -15,6 +15,11 @@ typedef struct ByteArray {
     int length;
 } ByteArray;
 
+typedef struct DataPtr {
```
See previous comment, seems like could just use the ByteArray type instead of adding an additional type.

vcabbage commented:
Removed the `DataPtr` struct and replaced `Mat_ToBytes` with `Mat_DataPtr` in `Mat.ToBytes`.

deadprogram commented:
@vcabbage this is really good stuff. Seems ready to merge now, right?

vcabbage commented:
Yep, should be all set. Thanks for reviewing!
